lib/journald/src/lib.rs

```diff
@@ -0,0 +1,113 @@
+#![allow(non_camel_case_types)]
+
+#[macro_use]
+extern crate dlopen_derive;
+
+use dlopen::wrapper::{Container, WrapperApi};
+use libc::size_t;
+use std::collections::HashMap;
+use std::io::{Error as IOError, Result as IOResult};
+use std::iter;
+use std::os::raw::c_int;
+use std::ptr::null_mut;
+
+pub const SD_JOURNAL_LOCAL_ONLY: c_int = 1;
+pub const SD_JOURNAL_RUNTIME_ONLY: c_int = 2;
+pub const SD_JOURNAL_SYSTEM: c_int = 4;
+pub const SD_JOURNAL_CURRENT_USER: c_int = 8;
+
+pub enum sd_journal {}
+
+#[derive(WrapperApi)]
+struct LibSystemd {
+    sd_journal_open: extern "C" fn(ret: *mut *mut sd_journal, flags: c_int) -> c_int,
+    sd_journal_close: extern "C" fn(j: *mut sd_journal),
+    sd_journal_next: extern "C" fn(j: *mut sd_journal) -> c_int,
+    sd_journal_seek_head: extern "C" fn(j: *mut sd_journal) -> c_int,
+    sd_journal_restart_data: extern "C" fn(j: *mut sd_journal),
+    sd_journal_enumerate_data:
+        extern "C" fn(j: *mut sd_journal, data: *const *mut u8, l: *mut size_t) -> c_int,
+}
+
+fn load_lib() -> Result<Container<LibSystemd>, dlopen::Error> {
+    unsafe { Container::load("libsystemd.so") }
+}
+
+pub struct Journal {
+    lib: Container<LibSystemd>,
+    journal: *mut sd_journal,
+}
+
+unsafe impl Send for Journal {}
+
+pub type Record = HashMap<String, String>;
+
+impl Journal {
+    pub fn open(flags: c_int) -> IOResult<Journal> {
+        // Each Journal structure gets their own handle to the library,
+        // but I couldn't figure out how to make lazy_static work.
+        let lib = load_lib().map_err(|err| IOError::new(std::io::ErrorKind::Other, err))?;
+
+        let mut journal = null_mut();
+        sd_result(lib.sd_journal_open(&mut journal, flags))?;
+        sd_result(lib.sd_journal_seek_head(journal))?;
+        Ok(Journal { lib, journal })
+    }
+
+    pub fn get_record(&mut self) -> IOResult<Record> {
```

Comment:
what is the difference between this one and the iterator? might be worth a doc comment

Reply:
I copied the naming from the `rust-systemd` crate (except that it doesn't have an iterator). `libsystemd` requires two steps to fetch the next record -- advance to the next record, and then get all its fields (one by one ugh). I'll add a comment and maybe rename it to clarify.